File: Core/TuningCore/KBM.swift
//
//  KBM.swift
//  Tenney
//
//  Created by Sebastian Suarez-Solis on 10/1/25.
//

import Foundation

/// Minimal .kbm mapping (common fields). Forgiving reader/writer.
struct KeyboardMapping: Equatable, Codable {
    var mappingSize: Int                 // typically 12
    var firstMIDINote: Int               // e.g., 0
    var lastMIDINote: Int                // e.g., 127
    var middleNote: Int                  // e.g., 69
    var referenceFrequencyHz: Double     // e.g., 440.0
    var referenceDegreeIndex: Int        // 0-based index in scale (usually 0)
    var degreeOfNote: [Int]              // length mappingSize, entries -1 for unmapped

    static func parse(_ text: String) throws -> KeyboardMapping {
        var lines = text.split(whereSeparator: \.isNewline).map(String.init)
        lines = lines.compactMap { line in
            let t = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !t.isEmpty else { return nil }
            if t.hasPrefix("!") { return nil }
            return t
        }
        func requireInt(_ idx: Int, _ name: String) throws -> Int {
            guard idx < lines.count, let v = Int(lines[idx].split(separator: " ").first ?? "") else {
                throw NSError(domain: "KBM", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing \(name)"])
            }
            return v
        }
        func requireDouble(_ idx: Int, _ name: String) throws -> Double {
            guard idx < lines.count, let v = Double(lines[idx].split(separator: " ").first ?? "") else {
                throw NSError(domain: "KBM", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing \(name)"])
            }
            return v
        }

        let mappingSize = try requireInt(0, "mapping size")
        let firstMIDINote = try requireInt(1, "first MIDI note")
        let lastMIDINote = try requireInt(2, "last MIDI note")
        let middleNote = try requireInt(3, "middle note")
        let refDegree = try requireInt(4, "reference degree")
        let refFreq = try requireDouble(5, "reference frequency")
        var degreeOfNote: [Int] = []
        var idx = 6
        for _ in 0..<mappingSize {
            if idx >= lines.count { degreeOfNote.append(-1) } else {
                degreeOfNote.append(Int(lines[idx].split(separator: " ").first ?? "-1") ?? -1)
            }
            idx += 1
        }

        return KeyboardMapping(
            mappingSize: mappingSize,
            firstMIDINote: firstMIDINote,
            lastMIDINote: lastMIDINote,
            middleNote: middleNote,
            referenceFrequencyHz: refFreq,
            referenceDegreeIndex: refDegree,
            degreeOfNote: degreeOfNote
        )
    }

    func serialize() -> String {
        var out: [String] = []
        out.append("! Tenney .kbm")
        out.append("\(mappingSize)")
        out.append("\(firstMIDINote)")
        out.append("\(lastMIDINote)")
        out.append("\(middleNote)")
        out.append("\(referenceDegreeIndex)")
        out.append(String(format: "%.6f", referenceFrequencyHz))
        for d in degreeOfNote { out.append("\(d)") }
        return out.joined(separator: "\n")
    }
}

File: apps/android/formats/src/main/kotlin/com/tenney/formats/KBM.kt
package com.tenney.formats

data class KeyboardMapping(
    val mappingSize: Int,
    val firstMIDINote: Int,
    val lastMIDINote: Int,
    val middleNote: Int,
    val referenceFrequencyHz: Double,
    val referenceDegreeIndex: Int,
    val degreeOfNote: List<Int>
) {
    companion object {
        fun parse(text: String): KeyboardMapping {
            var lines = text.split(Regex("\\R")).map { it.trim() }
            lines = lines.filter { it.isNotEmpty() && !it.startsWith("!") }

            fun requireInt(index: Int, name: String): Int {
                val token = lines.getOrNull(index)?.split(Regex("\\s+"))?.firstOrNull()
                return token?.toIntOrNull() ?: error("Missing $name")
            }

            fun requireDouble(index: Int, name: String): Double {
                val token = lines.getOrNull(index)?.split(Regex("\\s+"))?.firstOrNull()
                return token?.toDoubleOrNull() ?: error("Missing $name")
            }

            val mappingSize = requireInt(0, "mapping size")
            val firstMIDINote = requireInt(1, "first MIDI note")
            val lastMIDINote = requireInt(2, "last MIDI note")
            val middleNote = requireInt(3, "middle note")
            val referenceDegreeIndex = requireInt(4, "reference degree")
            val referenceFrequencyHz = requireDouble(5, "reference frequency")

            val degreeOfNote = mutableListOf<Int>()
            var idx = 6
            repeat(mappingSize) {
                val token = lines.getOrNull(idx)?.split(Regex("\\s+"))?.firstOrNull()
                degreeOfNote.add(token?.toIntOrNull() ?: -1)
                idx += 1
            }

            return KeyboardMapping(
                mappingSize = mappingSize,
                firstMIDINote = firstMIDINote,
                lastMIDINote = lastMIDINote,
                middleNote = middleNote,
                referenceFrequencyHz = referenceFrequencyHz,
                referenceDegreeIndex = referenceDegreeIndex,
                degreeOfNote = degreeOfNote
            )
        }
    }

    fun serialize(): String {
        val out = mutableListOf<String>()
        out.add("! Tenney .kbm")
        out.add(mappingSize.toString())
        out.add(firstMIDINote.toString())
        out.add(lastMIDINote.toString())
        out.add(middleNote.toString())
        out.add(referenceDegreeIndex.toString())
        out.add(String.format("%.6f", referenceFrequencyHz))
        degreeOfNote.forEach { out.add(it.toString()) }
        return out.joinToString("\n")
    }
}
